test/functional/p2p_addr_relay.py

```diff
@@ -207,6 +225,69 @@ def blocksonly_mode_tests(self):
 
         self.nodes[0].disconnect_p2ps()
 
+    def rate_limit_tests(self):
+
+        for contype, tokens, no_relay in [("outbound-full-relay", 1001, False), ("block-relay-only", 0, True), ("inbound", 1, False)]:
+            self.log.info('Test rate limiting of addr processing for %s peers' % contype)
+            self.stop_node(0)
+            os.remove(os.path.join(self.nodes[0].datadir, "regtest", "peers.dat"))
+            self.start_node(0, [])
+            self.mocktime = int(time.time())
+            self.nodes[0].setmocktime(self.mocktime)
+            if contype == "inbound":
+                peer = self.nodes[0].add_p2p_connection(AddrReceiver())
+            else:
+                peer = self.nodes[0].add_outbound_p2p_connection(AddrReceiver(), p2p_idx=0, connection_type=contype)
+
+            # Check that we start off with empty addrman
+            addr_count_0 = len(self.nodes[0].getnodeaddresses(0))
+            assert_equal(addr_count_0, 0)
+
+            # Send 600 addresses. For all but the block-relay-only peer this should result in at least 1 address.
+            peer.send_and_ping(self.setup_rand_addr_msg(600))
+            addr_count_1 = len(self.nodes[0].getnodeaddresses(0))
+            assert_greater_than_or_equal(tokens, addr_count_1)
+            assert_greater_than_or_equal(addr_count_0 + 600, addr_count_1)
+            assert_equal(addr_count_1 > addr_count_0, tokens > 0)
+
+            # Send 600 more addresses. For the outbound-full-relay peer (which sends a GETADDR, and thus will
+            # process up to 1001 incoming addresses), this means more entries will appear.
+            peer.send_and_ping(self.setup_rand_addr_msg(600))
+            addr_count_2 = len(self.nodes[0].getnodeaddresses(0))
+            assert_greater_than_or_equal(tokens, addr_count_2)
+            assert_greater_than_or_equal(addr_count_1 + 600, addr_count_2)
+            assert_equal(addr_count_2 > addr_count_1, tokens > 600)
+
+            # Send 10 more. As we reached the processing limit for all nodes, this should have no effect.
+            peer.send_and_ping(self.setup_rand_addr_msg(10))
+            addr_count_3 = len(self.nodes[0].getnodeaddresses(0))
+            assert_greater_than_or_equal(tokens, addr_count_3)
+            assert_equal(addr_count_2, addr_count_3)
+
+            # Advance the time by 100 seconds, permitting the processing of 10 more addresses. Send 200,
+            # but verify that no more than 10 are processed.
+            self.mocktime += 100
+            self.nodes[0].setmocktime(self.mocktime)
+            if not no_relay:
+                tokens += 10
```

Comment:
You reuse this value of 10 lower down. How about saving it as a variable to remove that repetition (and also to test that the no_relay node's addr count doesn't increase):

```diff
--- a/test/functional/p2p_addr_relay.py
+++ b/test/functional/p2p_addr_relay.py
@@ -268,23 +268,23 @@ class AddrTest(BitcoinTestFramework):
             # but verify that no more than 10 are processed.
             self.mocktime += 100
             self.nodes[0].setmocktime(self.mocktime)
-            if not no_relay:
-                tokens += 10
+            new_tokens = 0 if no_relay else 10
+            tokens += new_tokens
             peer.send_and_ping(self.setup_rand_addr_msg(200))
             addr_count_4 = len(self.nodes[0].getnodeaddresses(0))
             assert_greater_than_or_equal(tokens, addr_count_4)
-            assert_greater_than_or_equal(addr_count_3 + 10, addr_count_4)
+            assert_greater_than_or_equal(addr_count_3 + new_tokens, addr_count_4)
 
             # Advance the time by 1000 seconds, permitting the processing of 100 more addresses. Send 200,
             # but verify that no more than 100 are processed (and at least some).
             self.mocktime += 1000
             self.nodes[0].setmocktime(self.mocktime)
-            if not no_relay:
-                tokens += 100
+            new_tokens = 0 if no_relay else 100
+            tokens += new_tokens
             peer.send_and_ping(self.setup_rand_addr_msg(200))
             addr_count_5 = len(self.nodes[0].getnodeaddresses(0))
             assert_greater_than_or_equal(tokens, addr_count_5)
-            assert_greater_than_or_equal(addr_count_4 + 100, addr_count_5)
+            assert_greater_than_or_equal(addr_count_4 + new_tokens, addr_count_5)
             assert_equal(addr_count_5 > addr_count_4, not no_relay)
 
             self.nodes[0].disconnect_p2ps()
```

Reply:
Maybe have just one constant in the test "bucket size = 1000" and derive everything from it? So that if the bucket size is changed in the future in `net_processing.cpp` this test is easier to adjust.

Reply:
@jnewbery Done.

@vasild It's not entirely clear how to do that sufficiently generically (we'll probably want different test cases entirely if the rate limit changes).